apps/src/javalab/JavalabFileHelper.js

```diff
@@ -18,19 +18,68 @@ export function getDefaultFileContents(filename, viewMode) {
 export const getTabKey = index => `file-${index}`;
 
 export const fileMetadataForEditor = (sources, isEditingStartSources) => {
+  /*
+  'sources' contains file information as key-value pairs.
+  The key is the file name such as 'Class1.java'.
+  The value is a file object with the keys: text (source code),
+  tabOrder (order of the file tab in orderedTabs from 0 to (number of files - 1)),
+  isVisible, and isValidation (default is false)
+  The file tabKey names (such as 'file-0', 'file-1') are stored in fileMetadata.
+  For example, if sources contains the following:
+  {
+    'ClassOne.java': {text: '', tabOrder: 2, isVisible: true},
+    'ClassTwo.java': {text: '', tabOrder: 0, isVisible: true},
+    'ClassThree.java': {text: '', tabOrder: 1, isVisible: true}
+  },
+  then depending on how Object.keys iterates through the keys on sources, we could have
+  fileMetadata assigned: {'file-0': 'ClassOne.java', 'file-1': 'ClassTwo.java', 'file-2': 'ClassThree.java'}.
+  and orderedTabKeys would be ['file-1', 'file-2', 'file-0'].
+  */
   let fileMetadata = {};
   let orderedTabKeys = [];
-
-  Object.keys(sources).forEach(file => {
-    if (sources[file].isVisible || isEditingStartSources) {
-      // tabIndex should be the length of orderedTabKeys, not the index
-      // from the list, because we skip hidden files in the tabs.
-      const tabIndex = orderedTabKeys.length;
-      let tabKey = getTabKey(tabIndex);
-      fileMetadata[tabKey] = file;
-      orderedTabKeys.push(tabKey);
+  let unorderedTabKeys = [];
+  let fileIndex = 0;
+  let isValid = true;
+  Object.keys(sources).forEach(fileName => {
+    if (sources[fileName].isVisible || isEditingStartSources) {
+      let fileTabKey = getTabKey(fileIndex);
+      fileMetadata[fileTabKey] = fileName;
+      let tabOrder = sources[fileName].tabOrder;
+      // Files that are stored may not currently have an tabOrder assigned so that
+      // order is undefined
+      // Also check if two files have the same tabOrder. If so, then orderedTabKeys is invalid
+      if (
```

Comment:
nit: you could also check if `isValid` is already false here to short-circuit this condition